File: backend/internal/realtime/protocol.go
package realtime

import (
	"encoding/json"
	"time"
)

const ProtocolVersion = 1

type Event struct {
	Type            string          `json:"type"`
	ProtocolVersion int             `json:"protocol_version"`
	RequestID       string          `json:"request_id,omitempty"`
	ClientMsgID     string          `json:"client_msg_id,omitempty"`
	RoomID          string          `json:"room_id,omitempty"`
	Timestamp       string          `json:"timestamp,omitempty"`
	Payload         json.RawMessage `json:"payload,omitempty"`
}

type Participant struct {
	ParticipantID string `json:"participant_id"`
	DisplayName   string `json:"display_name"`
	Status        string `json:"status"`
}

func NewEvent(eventType string) Event {
	return Event{
		Type:            eventType,
		ProtocolVersion: ProtocolVersion,
		Timestamp:       time.Now().UTC().Format(time.RFC3339Nano),
	}
}

func Payload(value any) json.RawMessage {
	bytes, err := json.Marshal(value)
	if err != nil {
		return nil
	}
	return bytes
}

File: frontend/src/lib/realtime/protocol.ts
export const PROTOCOL_VERSION = 1;

export type Participant = {
  participant_id: string;
  display_name: string;
  status: "online" | "away" | "busy";
};

export type RealtimeEvent<T = unknown> = {
  type: string;
  protocol_version: number;
  request_id?: string;
  client_msg_id?: string;
  room_id?: string;
  timestamp?: string;
  payload?: T;
};

export function event<T>(type: string, payload?: T, options: Partial<RealtimeEvent> = {}): RealtimeEvent<T> {
  return {
    type,
    protocol_version: PROTOCOL_VERSION,
    request_id: options.request_id || crypto.randomUUID(),
    client_msg_id: options.client_msg_id,
    room_id: options.room_id,
    payload
  };
}
